lib/Doctrine/DBAL/Query/Expression/CompositeExpression.php

```diff
@@ -84,6 +88,39 @@ public function add($part)
         return $this;
     }
 
+    /**
+     * Returns a new CompositeExpression with the given parts added.
+     *
+     * @param self|string $part
+     * @param self|string ...$parts
+     */
+    public function with($part, ...$parts) : self
+    {
+        $that = clone $this;
+
+        if (! $this->isEmptySelf($part)) {
```

Comment:
I'm not sure if these checks need to be added to the new API. Consumers will opt-in to using it, so they will need to make sure they are not using empty composite expressions. Instead of sweetening the pill for them, we should make sure their code is ready for the upgrade.

Reply:
That's a good point. At some point, we want to disallow empty expressions. However, we can't do this in 2.x, so what should be do here?

- leave as is: skip empty expressions
- don't make this check: empty expressions will be added to the expression, and produce invalid SQL
- throw an exception?

Reply:
I'd just remove the check from here. It's already stated in the code documentation that by the compilation time an expression cannot be empty:

https://github.com/doctrine/dbal/blob/d191e953778c24efb0fd4bd3a4752e443563239e/lib/Doctrine/DBAL/Query/Expression/ExpressionBuilder.php#L80-L81

Reply:
Makes sense: no sane code base should ever contain an empty expression anyway, as this would produce invalid SQL. Done in 1341e462da44e4d5734c96d765ef8008029ca2f7.